proof/refine/ARM/Schedule_R.thy

```diff
@@ -3986,6 +3986,236 @@ lemma refillPopHead_valid_refills'[wp]:
+   \<lbrace>\<lambda>_ s. P (scs_of2 s)\<rbrace>"
+  unfolding merge_refills_def
+  apply (wpsimp simp: update_refill_hd_rewrite refill_pop_head_def
+                  wp: get_refills_wp set_refills_wp update_sched_context_wp)
+  by (clarsimp simp: is_active_sc2_def obj_at_def opt_map_red)
+
+(* if the loop guard is true, the refill length is greater than one *)
+lemma mergeRefills_corres:
+  "corres dc (sc_at sc_ptr and is_active_sc2 sc_ptr and pspace_aligned and pspace_distinct
+              and (\<lambda>s. ((\<lambda>sc. 1 < length (sc_refills sc)) |< scs_of2 s) sc_ptr))
+             (valid_refills' sc_ptr)
+                 (merge_refills sc_ptr) (mergeRefills sc_ptr)"
+  unfolding mergeRefills_def merge_refills_def merge_refill_def
+  apply (rule_tac Q="is_active_sc' sc_ptr" in corres_cross_add_guard)
+   apply (fastforce dest!: is_active_sc'_cross[OF state_relation_pspace_relation])
+  apply (rule corres_cross[where Q' = "sc_at' sc_ptr", OF sc_at'_cross_rel], fastforce)
+  apply (rule_tac Q="\<lambda>s'. ((\<lambda>sc'. 1 < scRefillCount sc') |< scs_of' s') sc_ptr" in corres_cross_add_guard)
+   apply clarsimp
+   apply (drule (2) state_relation_sc_relation)
+   apply (clarsimp simp: sc_relation_def)
+   apply (clarsimp simp: valid_refills'_def obj_at_simps is_sc_obj opt_map_red)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF refillPopHead_corres])
+      apply (rule updateRefillHd_corres, simp)
+       apply (clarsimp simp: refill_map_def merge_refill_def)+
+     apply (rule hoare_strengthen_post[where Q="\<lambda>_. sc_at sc_ptr and active_sc_at sc_ptr", rotated])
+      apply (simp add: active_sc_at_equiv is_active_sc_rewrite[symmetric])
+     apply (wpsimp wp: refill_pop_head_sc_active)
+    apply (wpsimp wp: refillPopHead_scs_of')
+   apply (clarsimp simp: obj_at_def is_sc_obj opt_map_red is_active_sc_rewrite active_sc_at_equiv
+                         sc_refills_sc_at_rewrite)
+  apply (fastforce simp: valid_refills'_def obj_at_simps opt_map_red refillNextIndex_def)
+  done
+
+crunches mergeRefills
+  for sc_at'[wp]: "\<lambda>s. P (sc_at' p s)"
+  (simp: objBits_simps)
+
+lemma mergeRefills_valid_refills'[wp]:
+  "\<lbrace>(\<lambda>s. ((\<lambda>sc. 1 < scRefillCount sc) |< scs_of' s) scp) and valid_refills' scp\<rbrace>
+   mergeRefills p
+   \<lbrace>\<lambda>_. valid_refills' scp\<rbrace>"
+  unfolding mergeRefills_def
+  apply (wpsimp simp: updateRefillHd_def refillPopHead_def wp: updateSchedContext_wp)
+  by (fastforce simp: valid_refills'_def obj_at_simps ps_clear_upd refillNextIndex_def opt_map_red)
+
+lemma no_fail_refillPopHead[wp]:
+  "no_fail (sc_at' scp and (\<lambda>s'. ((\<lambda>sc'. 0 < scRefillCount sc') |< scs_of' s') scp)) (refillPopHead scp)"
+  unfolding refillPopHead_def
+  by (wpsimp simp: obj_at_simps scBits_simps opt_map_red)
+
+lemma no_fail_mergeRefills[wp]:
+  "no_fail (sc_at' scp and (\<lambda>s'. ((\<lambda>sc'. 0 < scRefillCount sc') |< scs_of' s') scp)) (mergeRefills scp)"
```

Comment:
It's not really worth changing, but I'm a bit curious whether `crunch` would solve this and generate the same precondition.
If you're also curious it would be
```
crunches mergeRefills
  for (no_fail) no_fail[wp]:
  (needed wp and simp rules)
```

Reply:
It does work and it told me that we don’t need the second (big) precondition, which was nice :)